// @tanstack/start-plugin-core/dist/esm/vite/serialization-adapters-plugin.js
import { START_ENVIRONMENT_NAMES } from "../constants.js";
import { resolveViteId } from "../utils.js";
import { EMPTY_SERIALIZATION_ADAPTERS_MODULE, generateSerializationAdaptersModule } from "../serialization-adapters-module.js";
import { VIRTUAL_MODULES } from "@tanstack/start-server-core";
//#region src/vite/serialization-adapters-plugin.ts
var resolvedModuleId = resolveViteId(VIRTUAL_MODULES.pluginAdapters.replace("#", "%23"));
function escapeRegex(str) {
	return str.replace(/[.*+?^${}()|[\]\\#]/g, "\\$&");
}
function serializationAdaptersPlugin(opts) {
	return {
		name: "tanstack-start:plugin-adapters",
		enforce: "pre",
		resolveId: {
			filter: { id: new RegExp(escapeRegex(VIRTUAL_MODULES.pluginAdapters)) },
			handler(id) {
				if (id === VIRTUAL_MODULES.pluginAdapters) return resolvedModuleId;
			}
		},
		load: {
			filter: { id: new RegExp(escapeRegex(resolvedModuleId)) },
			handler(id) {
				if (id !== resolvedModuleId) return;
				const adapters = opts.adapters;
				if (!adapters || adapters.length === 0) return EMPTY_SERIALIZATION_ADAPTERS_MODULE;
				if (this.environment.name === START_ENVIRONMENT_NAMES.client) return generateSerializationAdaptersModule({
					adapters,
					runtime: "client"
				});
				if (this.environment.name === START_ENVIRONMENT_NAMES.server) return generateSerializationAdaptersModule({
					adapters,
					runtime: "server"
				});
				return EMPTY_SERIALIZATION_ADAPTERS_MODULE;
			}
		}
	};
}
//#endregion
export { serializationAdaptersPlugin };
