// @stackbit/cms-core/dist/content-store-utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getContentSourceIdForContentSource = getContentSourceIdForContentSource;
exports.getContentSourceId = getContentSourceId;
exports.getObjectId = getObjectId;
exports.getContentSourceDataByTypeAndProjectIdOrThrow = getContentSourceDataByTypeAndProjectIdOrThrow;
exports.getContentSourceDataByIdOrThrow = getContentSourceDataByIdOrThrow;
exports.findContentSourcesDataForTypeOrId = findContentSourcesDataForTypeOrId;
exports.getUserContextForSrcTypeThunk = getUserContextForSrcTypeThunk;
exports.getUserContextForSrcType = getUserContextForSrcType;
exports.isDocumentFieldOneOfFieldTypes = isDocumentFieldOneOfFieldTypes;
exports.getDocumentFieldForLocale = getDocumentFieldForLocale;
exports.groupModelsByContentSource = groupModelsByContentSource;
exports.groupDocumentsByContentSource = groupDocumentsByContentSource;
exports.getCSIDocumentsAndAssetsFromContentSourceDataByIds = getCSIDocumentsAndAssetsFromContentSourceDataByIds;
exports.updateOperationValueFieldWithCrossReference = updateOperationValueFieldWithCrossReference;
exports.isContentChangesEmpty = isContentChangesEmpty;
exports.isContentChangeResultEmpty = isContentChangeResultEmpty;
exports.contentChangeResultCounts = contentChangeResultCounts;
exports.getErrorAtLine = getErrorAtLine;
exports.getCacheDir = getCacheDir;
const lodash_1 = __importDefault(require("lodash"));
const path_1 = __importDefault(require("path"));
const utils_1 = require("@stackbit/utils");
const types_1 = require("@stackbit/types");
function getContentSourceIdForContentSource(contentSource) {
    return getContentSourceId(contentSource.getContentSourceType(), contentSource.getProjectId());
}
function getContentSourceId(contentSourceType, srcProjectId) {
    return contentSourceType + ':' + srcProjectId;
}
function getObjectId(srcType, srcProjectId, srcObjectId) {
    return `${srcType}:${srcProjectId}:${srcObjectId}`;
}
function getContentSourceDataByTypeAndProjectIdOrThrow(srcType, srcProjectId, contentSourceDataById) {
    const contentSourceId = getContentSourceId(srcType, srcProjectId);
    return getContentSourceDataByIdOrThrow(contentSourceId, contentSourceDataById);
}
function getContentSourceDataByIdOrThrow(contentSourceId, contentSourceDataById) {
    const contentSourceData = contentSourceDataById[contentSourceId];
    if (!contentSourceData) {
        throw new Error(`Content source not found: '${contentSourceId}'.`);
    }
    return contentSourceData;
}
function findContentSourcesDataForTypeOrId({ contentSourceDataById, srcType, srcProjectId }) {
    return lodash_1.default.filter(contentSourceDataById, (contentSourceData) => {
        const srcTypeMatch = !srcType || contentSourceData.srcType === srcType;
        const srcProjectIdMatch = !srcProjectId || contentSourceData.srcProjectId === srcProjectId;
        return srcTypeMatch && srcProjectIdMatch;
    });
}
function getUserContextForSrcTypeThunk(user) {
    return (srcType) => getUserContextForSrcType(srcType, user);
}
function getUserContextForSrcType(srcType, user) {
    if (!user) {
        return undefined;
    }
    const connection = user?.connections?.find((connection) => connection.type === srcType);
    return {
        email: user.email,
        name: user.name,
        sso: user.sso,
        role: user.role,
        ...connection
    };
}
function isDocumentFieldOneOfFieldTypes(documentField, fieldTypes) {
    return fieldTypes.includes(documentField.type);
}
function getDocumentFieldForLocale(docField, locale) {
    if (docField && docField.localized) {
        if (!locale) {
            return null;
        }
        const { localized, locales, ...base } = docField;
        const localizedField = locales?.[locale];
        if (!localizedField) {
            return null;
        }
        return {
            ...base,
            ...localizedField
        };
    }
    else {
        return docField;
    }
}
function groupModelsByContentSource({ models }) {
    const modelMapByContentSource = {};
    for (const model of models) {
        const { srcType, srcProjectId, ...rest } = model;
        lodash_1.default.set(modelMapByContentSource, [srcType, srcProjectId, model.name], rest);
    }
    return modelMapByContentSource;
}
function groupDocumentsByContentSource({ documents }) {
    const documentMapByContentSource = {};
    for (const document of documents) {
        const { srcType, srcProjectId, ...rest } = document;
        (0, utils_1.append)(documentMapByContentSource, [srcType, srcProjectId], rest);
    }
    return documentMapByContentSource;
}
function getCSIDocumentsAndAssetsFromContentSourceDataByIds(contentSourceData, objects) {
    const documents = [];
    const assets = [];
    for (const object of objects) {
        if (object.srcObjectId in contentSourceData.csiDocumentMap) {
            documents.push(contentSourceData.csiDocumentMap[object.srcObjectId]);
        }
        else if (object.srcObjectId in contentSourceData.csiAssetMap) {
            assets.push(contentSourceData.csiAssetMap[object.srcObjectId]);
        }
    }
    return {
        documents,
        assets
    };
}
function updateOperationValueFieldWithCrossReference(type, refObject) {
    if ((0, types_1.isOneOfFieldTypes)(type, ['json', 'cross-reference'])) {
        return {
            type,
            value: refObject
        };
    }
    return {
        type,
        value: JSON.stringify(refObject)
    };
}
function isContentChangesEmpty(contentChanges) {
    return (!contentChanges ||
        (!contentChanges.documents?.length &&
            !contentChanges.assets?.length &&
            !contentChanges.scheduledActions?.length &&
            !contentChanges.deletedDocumentIds?.length &&
            !contentChanges.deletedAssetIds?.length &&
            !contentChanges.deletedScheduledActionIds?.length));
}
function isContentChangeResultEmpty(contentChangeResult) {
    return (contentChangeResult.createdDocuments.length === 0 &&
        contentChangeResult.createdAssets.length === 0 &&
        contentChangeResult.createdScheduledActions.length === 0 &&
        contentChangeResult.updatedDocuments.length === 0 &&
        contentChangeResult.updatedAssets.length === 0 &&
        contentChangeResult.updatedScheduledActions.length === 0 &&
        contentChangeResult.deletedDocuments.length === 0 &&
        contentChangeResult.deletedAssets.length === 0 &&
        contentChangeResult.deletedScheduledActions.length === 0);
}
function contentChangeResultCounts(contentChangeResult) {
    return {
        createdDocumentsCount: contentChangeResult.createdDocuments.length,
        createdAssetsCount: contentChangeResult.createdAssets.length,
        createdScheduledActionsCount: contentChangeResult.createdScheduledActions.length,
        updatedDocumentsCount: contentChangeResult.updatedDocuments.length,
        updatedAssetsCount: contentChangeResult.updatedAssets.length,
        updatedScheduledActionsCount: contentChangeResult.updatedScheduledActions.length,
        deletedDocumentsCount: contentChangeResult.deletedDocuments.length,
        deletedAssetsCount: contentChangeResult.deletedAssets.length,
        deletedScheduledActionsCount: contentChangeResult.deletedScheduledActions.length
    };
}
// eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
function getErrorAtLine(dropLastCallStackEntries, dropCallStackFunc) {
    const tempErr = new Error();
    Error.captureStackTrace(tempErr, dropCallStackFunc ?? getErrorAtLine);
    const atLine = tempErr.stack?.split('\n')[1 + dropLastCallStackEntries]?.trim();
    return atLine ? ` ${atLine}` : '';
}
function getCacheDir(stackbitConfig) {
    return stackbitConfig.cacheDir ?? path_1.default.resolve(stackbitConfig.dirPath, '.stackbit/cache');
}
