// @stackbit/types/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadJSON = loadJSON;
exports.getInterfaceVersion = getInterfaceVersion;
exports.getVersion = getVersion;
exports.isLocalizedField = isLocalizedField;
exports.getLocalizedFieldForLocale = getLocalizedFieldForLocale;
exports.isOneOfFieldTypes = isOneOfFieldTypes;
exports.isModelFieldOneOfFieldTypes = isModelFieldOneOfFieldTypes;
exports.isModelFieldSpecificPropsOneOfFieldTypes = isModelFieldSpecificPropsOneOfFieldTypes;
exports.isDocumentFieldOneOfFieldTypes = isDocumentFieldOneOfFieldTypes;
exports.defineStackbitConfig = defineStackbitConfig;
async function loadJSON(filePath) {
    // The `@stackbit/pacakge` can be imported in browser environment. To use
    // the types in client-side's code, or when using the
    // `getLocalizedFieldForLocale` utility method.
    // Therefore, don't use `require('fs/promises')`, as bundlers will try to
    // bundle the inline `fs/promises` and browser will not be able to load it.
    const module = 'fs/promises';
    const fs = require(module);
    return JSON.parse(await fs.readFile(filePath, {
        encoding: `utf8`
    }));
}
async function getInterfaceVersion() {
    const path = require('path');
    const packageJson = await loadJSON(path.join(__dirname, '../package.json'));
    return packageJson.version;
}
async function getVersion(options = {}) {
    const interfaceVersion = await getInterfaceVersion();
    if (options.contentSourceVersion) {
        return {
            interfaceVersion: interfaceVersion,
            contentSourceVersion: options.contentSourceVersion
        };
    }
    else if (options.packageJsonPath) {
        const packageJson = await loadJSON(options.packageJsonPath);
        return {
            interfaceVersion: interfaceVersion,
            contentSourceVersion: packageJson.version
        };
    }
    else {
        return {
            interfaceVersion: interfaceVersion,
            contentSourceVersion: ''
        };
    }
}
function isLocalizedField(field) {
    return !!field.localized;
}
/**
 * Returns non-localized version of a DocumentField for the provided `locale`.
 * If the `field` is localized, and the provided locale is not present, this
 * function returns `undefined`. If the `field` is not localized, this function
 * returns the `field` as-is ignoring the `locale`.
 *
 * @example Passing localized field
 * getLocalizedFieldForLocale({
 *   type: 'string',
 *   localized: true,
 *   locales: {
 *       en-US: { value: 'Hello' },
 *       es-ES: { value: 'Hola' },
 *       fr-FR: { value: 'Bonjour' }
 *   }
 * }, 'es-ES');
 * returns:
 * {
 *   type: 'string',
 *   value: 'Hola'
 * }
 *
 * @example Passing non-localized field
 * getLocalizedFieldForLocale({
 *   type: 'string',
 *   value: 'Hello'
 * });
 * returns:
 * {
 *   type: 'string',
 *   value: 'Hello'
 * }
 *
 * @param field
 * @param locale
 */
function getLocalizedFieldForLocale(field, locale) {
    if (field && field.localized) {
        if (!locale) {
            return undefined;
        }
        const { localized, locales, ...base } = field;
        const localizedField = locales?.[locale];
        if (!localizedField) {
            return undefined;
        }
        return {
            ...base,
            ...localizedField
        };
    }
    return field;
}
function isOneOfFieldTypes(fieldType, fieldTypes) {
    return fieldTypes.includes(fieldType);
}
function isModelFieldOneOfFieldTypes(modelField, fieldTypes) {
    return fieldTypes.includes(modelField.type);
}
function isModelFieldSpecificPropsOneOfFieldTypes(modelField, fieldTypes) {
    return fieldTypes.includes(modelField.type);
}
function isDocumentFieldOneOfFieldTypes(documentField, fieldTypes) {
    return fieldTypes.includes(documentField.type);
}
function defineStackbitConfig(stackbitConfig) {
    return stackbitConfig;
}
